mountpoint-s3/src/prefetch/parquet_stream.rs

```diff
@@ -213,32 +330,67 @@ where
+            let (cache, lru_cache) = cache_guard.get_or_insert_with(|| {
+                (HashMap::new(), AsyncRwLock::new(LruCache::new(1000 * 1024 * 1024)))
+                // 1 GB limit
+            });
 
             for part in parts {
                 let part_range = part.offset()..part.offset() + part.len() as u64;
-                trace!("Pushing part to queue from client: {:?}", part_range);
+                // list of row group cols and the ranges they have
+                let mut ranges = Vec::new();
+                for (row_group_col, map) in cache.iter() {
+                    ranges.push((*row_group_col, map.keys()));
+                }
+
+                trace!(
+                    "Pushing part to queue from client: {:?} where cache was {:?}",
+                    part_range,
+                    ranges
+                );
+
                 part_queue_producer.push(Ok(part.clone()));
 
-                let split_index: usize = cols.partition_point(|(_, col_range)| col_range.start < part_range.end);
-                cols.truncate(split_index);
+                // Filters out columns so we need to walk through less columns on the next part (next round) leveraging non-overlapping, structured order
+                let mut new_cols = Vec::new();
+                let mut remaining_cols = Vec::new();
 
-                for ((row_group, column), _) in cols.iter() {
-                    let col_cache = rowgroup_col_cache
-                        .as_mut()
-                        .unwrap()
-                        .entry((*row_group, *column))
-                        .or_insert_with(BTreeMap::new);
+                for (row_group_col, col_range) in cols.iter() {
+                    if col_range.start < part_range.end && part_range.start < col_range.end {
+                        new_cols.push((*row_group_col, col_range.clone()));
+                        if col_range.end > part_range.end {
+                            remaining_cols.push((*row_group_col, col_range.clone()));
+                        }
+                    } else if col_range.start >= part_range.end {
+                        remaining_cols.push((*row_group_col, col_range.clone()));
+                    }
+                }
 
-                    if let Err(e) = merge_ranges(col_cache, part_range.clone(), part.get_checksummed_bytes().clone()) {
-                        error!("Error merging ranges: {:?}", e);
+                *cols = remaining_cols;
+
+                for (row_group_col, _col_range) in new_cols {
+                    let col_cache = cache.entry(row_group_col).or_insert_with(BTreeMap::new);
+                    if let Err(e) = merge_ranges(
+                        col_cache,
+                        part_range.clone(),
+                        part.get_checksummed_bytes().clone(),
+                        metadata
+                            .as_ref()
+                            .unwrap()
```

Comment:
If we can safely `unwrap` here, why are we passing an `Option` in the first place?